### PR TITLE
Fixed the custom query not being handled by interface's relations

### PR DESCRIPTION
**Pull request by EdwinDayot**
Fixes the #421 

An `Interface` requires a `public function types` returning an array of `GraphQL::type()` if it contains a relation with a custom `query`

Still, I'm not a huge fan of this method, so every new idea is appreciated!

### CODE REVIEW

**Comment by EdwinDayot** [comment]
Found out that interfaces public function types were useless

**Comment by mfn** [comment]
Phu, the changes seem quite complex… I mean, on one side it looks like you just moved stuff around into a dedicated method, but hard to say.

Btw, this one test is failing; what is your test expectation here?

I'm a bit unsure if this change / added complexity is worth it, in its current form (no better suggestion though yet).

**Comment by EdwinDayot** [comment]
The changes are pretty simple actually. Instead of executing all the code before any database request, in the case of an interface's relation, it is executed on the relation's query, as we don't know what model will be requested. Maybe this solution is better than nothing, and if anyone complains about the complexity, or anything, maybe someone will PR something else?

Anyway, if anyone has something better in mind...

**Comment by mfn** [comment]
Ok, thx for the info.

I will take a closer look for the review, but that might take a while. If you want to add more details to the source changes (i.e. via github inline comments) to guide the reviews, that would be much appreciated.

File: src/Support/SelectFields.php [hunk 1]
```diff
@@ -180,49 +188,17 @@ protected static function handleFields(
...
-                        }
-                        // If 'HasMany', then add it in the 'with'
-                        elseif ((is_a($relation, HasMany::class) || is_a($relation, MorphMany::class) || is_a($relation, HasOne::class) || is_a($relation, MorphOne::class))
-                            && ! array_key_exists($foreignKey, $field)) {
-                            $segments = explode('.', $foreignKey);
-                            $foreignKey = end($segments);
-                            if (! array_key_exists($foreignKey, $field)) {
-                                $field['fields'][$foreignKey] = self::FOREIGN_KEY;
-                            }
-                        }
```
Comment (EdwinDayot) [inline src/Support/SelectFields.php]: Moved to `addForeignKey`

File: src/Support/InterfaceType.php [hunk 2]
```diff
@@ -39,6 +54,11 @@ public function getAttributes(): array
             $attributes['resolveType'] = $resolver;
         }
 
+        $resolver = $this->getTypesResolver();
+        if ($resolver) {
+            $attributes['types'] = $resolver;
+        }
+
```
Comment (EdwinDayot) [inline src/Support/InterfaceType.php]: We need to add a method in each `Interface`: `types` returning an array of `Type`s. This way, we know what `Type`s to expect while executing our relation's query.

File: src/Support/SelectFields.php [hunk 3]
```diff
@@ -390,6 +366,116 @@ private static function isMongodbInstance(GraphqlType $parentType): bool
...
+
+            self::addAlwaysFields($fieldObject, $field, $parentTable, true);
+
+            if ($newParentType instanceof WrappingType) {
+                $newParentType = $newParentType->getWrappedType(true);
+            }
+
+            if (isset($newParentType->config['types'])) {
+                foreach ($newParentType->config['types']() as $type) {
+                    if (app($type->config['model'])->getTable() === $query->getRelated()->getTable()) {
```
Comment (EdwinDayot) [inline src/Support/SelectFields.php]: For each types of the interface, we are fetching the table of the model. If it is the same table as the one we are currently querying from, then we should build our custom query, and execute it. (`return $callable($query);`)

File: src/Support/SelectFields.php [hunk 4]
```diff
@@ -390,6 +366,116 @@ private static function isMongodbInstance(GraphqlType $parentType): bool
...
+                        $callable = self::getSelectableFieldsAndRelations($queryArgs, $field, $type, $customQuery, false, $ctx, false);
+
+                        return $callable($query);
+                    }
+                }
+            }
+            /** @var callable $callable */
+            $callable = self::getSelectableFieldsAndRelations($queryArgs, $field, $newParentType, $customQuery, false, $ctx, false);
+
+            return $callable($query);
```
Comment (EdwinDayot) [inline src/Support/SelectFields.php]: Otherwise, we should call the custom query without giving it the relation's type

File: tests/Database/SelectFields/InterfaceTests/InterfaceTest.php [hunk 5]
```diff
@@ -78,7 +78,7 @@ public function testGeneratedRelationSqlQuery(): void
 
         $this->assertSqlQueries(<<<'SQL'
 select "id", "title" from "posts";
-select * from "comments" where "comments"."post_id" = ? and "comments"."post_id" is not null order by "comments"."id" asc;
+select "comments"."title", "comments"."post_id" from "comments" where "comments"."post_id" in (?) and "id" >= ? order by "comments"."id" asc;
```
Comment (EdwinDayot) [inline tests/Database/SelectFields/InterfaceTests/InterfaceTest.php]: It results in the fact that the interface gives the fields that should be selected for the relation instead of selecting `*`, and it allows a custom query to be passed to this relation's query.

File: src/Support/SelectFields.php [hunk 6]
```diff
@@ -390,6 +366,116 @@ private static function isMongodbInstance(GraphqlType $parentType): bool
...
+        array &$with,
+        $ctx,
+        $fieldObject,
+        $key,
+        $customQuery
+    ) {
+        $relationsKey = Arr::get($fieldObject->config, 'alias', $key);
+
+        $with[$relationsKey] = function ($query) use ($queryArgs, $field, $parentType, &$select, $ctx, $customQuery, $key, $fieldObject) {
+            $parentTable = self::isMongodbInstance($parentType) ? null : self::getTableNameFromParentType($parentType);
```
Comment (EdwinDayot) [inline src/Support/SelectFields.php]: This part is executed when a `Query` calls the `->with()` method of the model.

See it as a query like so:
```php
MyModel::with(function ($query) {
  // We are here
})->get();
```

**Comment by mfn** [comment]
Thanks for your effort _and_ your patience so far, really appreciated.

I took a longer look today and, whilst not easy, the approach seems 👍 to me!

But my feeling is that it's lacking tests. Only one existing test adapt (which I remember was added in advance for exactly this moment) isn't much.
`\Rebing\GraphQL\Support\SelectFields::handleInterfaceFields` also seems to have some uncovered parts in the area of `if (isset($newParentType->config['types'])) {`.

My suggestion would be these steps in order:
- please add much tests as you can think of / possible
- invite others to also give feedback; currently thinking about @crissi as being also quite active lately
- final review and 🤞 merge

How does that sound for your?

PS: without getting ahead, `handleInterfaceFields` seems like it could use more specific PHP types on the function arguments?

Thanks!

**Comment by mfn** [comment]
@EdwinDayot I see so often that you've to make micro commits for the StyleCI violations: it should be _rather_ easy => just click on the "Details" and download / apply the patch.

Good news: we want to switch to something else which allows devs to simply run php cs fixer locally, but it's not yet done => https://github.com/rebing/graphql-laravel/pull/502

**Comment by EdwinDayot** [comment]
Could not find the issue with Laravel 5.5 for the moment, if anyone has an idea...

**Comment by EdwinDayot** [comment]
> @EdwinDayot I see so often that you've to make micro commits for the StyleCI violations: it should be _rather_ easy => just click on the "Details" and download / apply the patch.
> 
> Good news: we want to switch to something else which allows devs to simply run php cs fixer locally, but it's not yet done => #502

Yeah, that's what I am doing. I wasn't aware of this feature at the beginning, but now, I'm using it :)

**Comment by EdwinDayot** [comment]
Help wanted

**Comment by mfn** [comment]
Ops, the StyleCI integration is still active. On it!

File: tests/Database/SelectFields/InterfaceTests/LikableInterfaceType.php [hunk 7]
```diff
@@ -24,7 +33,12 @@ public function fields(): array
             ],
             'title' => [
                 'type' => Type::nonNull(Type::string()),
-                'alias' => 'title',
+            ],
+            'likes' => [
+                'type' => Type::listOf(GraphQL::type('Like')),
+                'query' => function (array $args, MorphMany $query) {
+                    return $query->whereRaw('1=1');
```
Comment (mfn) [inline tests/Database/SelectFields/InterfaceTests/LikableInterfaceType.php]: This code is never executed => is this expected?

**Comment by mfn** [comment]
> Help wanted

I believe there's single "fix" for this problem: Laravel 5.5 simply generates a different query and I've adapted to test to consider a different assertion 🤷‍♀️

Tests are all green now (ignore StyleCI…)

Is this PR ready for review?

I just noticed _a lot_ of whitespace changes on phpdoc in `SelectFields` and many other kinds of formatting changes => can you please undo this and make "as clean PR as possible"?

File: tests/Database/SelectFields/InterfaceTests/LikableInterfaceType.php (hunk 7 above)
Comment (EdwinDayot) [reply tests/Database/SelectFields/InterfaceTests/LikableInterfaceType.php]: It is now resolved, I added a new test

**Comment by EdwinDayot** [comment]
The PHPDoc has been fixed

File: tests/Database/SelectFields/InterfaceTests/CommentType.php [hunk 8]
```diff
@@ -20,12 +19,7 @@ public function fields(): array
     {
         $interface = GraphQL::type('LikableInterface');
 
-        return [
-                'title' => [
-                    'type' => Type::nonNull(Type::string()),
-                    'alias' => 'title',
-                ],
-            ] + $interface->getFields();
+        return $interface->getFields();
```
Comment (mfn) [inline tests/Database/SelectFields/InterfaceTests/CommentType.php]: Is this change actually related somehow or just a cleanup?

Comment (EdwinDayot) [reply tests/Database/SelectFields/InterfaceTests/CommentType.php]: Just some cleanup :)

**Comment by EdwinDayot** [comment]
Readme sounds good to me

**Comment by mfn** [comment]
Preparing the merge now

**Comment by mfn** [comment]
Had to restart the travis jobs, thus it took longer. They were all green the first time, but the feedback loop to github didn't seem to work 🤷‍♀️ Retriggering the jobs on travis fixed it 👍 

Thanks for the PR :)

**Comment by EdwinDayot** [comment]
Nice :)
